### DISCUSSION TITLE
Multispeaker development progress

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#156] [Archived]
<USER>>>> georroussos<br/> <DATE> [March 6, 2020, 3:18pm]<br/> <br/> After some email discourse with Eren, I am creating this thread for<br/> multi-speaker related progress on the Mozilla TTS. Particular objectives<br/> include manipulating speaker encoder embeddings for purposes of voice<br/> altering (creation of mixed voices) and conditioning the TTS using them.<br/> <br/> As Eren mentioned to me, there is no multi-speaker doc on the git at<br/> present. My goal is to maybe write up something myself, if I come up<br/> with anything worthy of looking into.<br/> <br/> If anyone has any questions/results/thoughts they want to share, please!<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/multispeaker-development-progress]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 12, 2020, 1:37pm]<br/> <br/> As I have been discussing with Eren, I am interested in seeing how we<br/> can manipulate a TTS voice quality, using external embeddings, however I<br/> have been having trouble using external embeddings extracted using the<br/> speaker encoder. Eren suggested initializing the speaker embedding layer<br/> with the tensor I have and then freezing the weights, however I have not<br/> really been successful. If anyone has ever worked with it or has any<br/> guidance, please go ahead!<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 17, 2020, 10:03am]<br/> <br/> Hi again everyone,<br/> <br/> A little update. I was able to, after all, modify the pipeline a bit and<br/> feed it my own custom embeddings. Some info on it: I extracted them<br/> using the speaker encoder. They were the mean of embeddings coming from<br/> different speakers (from LibriTTS). I modified Tacotron2 to feed these<br/> embeddings during training and inference time and tried to load a<br/> pretrained, on LJSpeech, model and see if the voice would change in any<br/> way when the embeddings were used. The voice did not change, but the<br/> prosody does change, when the embeddings are used (attaching 2 samples).<br/> I suspect it is due to the nature of the model being single-speaker, so<br/> it can simply not generalize, however if anyone has any input, let's<br/> hear it.<br/> <br/> Now I will try to train a multi-speaker model on Tacotron and see if I<br/> can feed it my own embeddings and maybe generate novel speakers. Very<br/> interesting to see results.<br/> <br/> any thoughts? Maybe I did<br/> something wrong, but I did hope maybe finetuning the pretrained model on<br/> the embeddings for a little while might change the voice, but alas, it<br/> did not...<br/> <br/> [Archive.zip](https://discourse.mozilla.org/uploads/short-url/2vRL5BHSrZGn1mCAaDC1MwGaWsD.zip)<br/> (199.2 KB)<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [March 17, 2020, 11:35am]<br/> <br/> you need to train the model in multi-speaker settings to see it works<br/> with the embeddings. Single speaker model would not generalize for a<br/> different speaker.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 17, 2020, 11:44am]<br/> <br/> I thought the same. Well then, it is what I shall do. Definitely<br/> promising that, even in single speaker, custom embeddings change the<br/> prosody and I think it is a good sign for GST modelling. Will come back<br/> with multi-speaker results.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 18, 2020, 11:54am]<br/> <br/> At what amount of steps should I start to notice eval alignment<br/> increase? I also wonder if a batch size of 64 is a good idea on 4 GPUS.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [March 18, 2020, 12:31pm]<br/> <br/> multi GPU training does not improve the converge speed for TTS if you<br/> dont tune the hyperparameters like the learning rate<br/> <br/> It depends on your dataset but you should see somethings after a day.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 18, 2020, 12:47pm]<br/> <br/> I was training on the 100 variant of LibriTTS, with a learning rate of<br/> 0.0001 and forward attention, a batch size of 64 and original prenet;<br/> for about a day. However, it seemed to overfit and so I am now trying<br/> bidirectional decoder; however, with a batch size of 64 it kept crapping<br/> out, hence why I asked. I am not trying with a size of 32 and bn prenet.<br/> Would you have any recommendations?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [March 18, 2020, 12:48pm]<br/> <br/> Not really. Just see what works for you.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 18, 2020, 12:52pm]<br/> <br/> Thanks. Will keep trying things and see what works. Very exciting to<br/> play with the power of embeddings.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 20, 2020, 11:36am]<br/> <br/> Hi again, been trying to train this whole week, but unfortunately<br/> everything has been overfitting (plateuing early on and no improvements<br/> after a day). I have been training on Tacotron, first with LibriTTS 100,<br/> and then LibriTTS 360, female only. I tried Forward Attention at first<br/> (all three switches in config enabled), and bidirectional at some point,<br/> but unfortunately it craps out because of memory. I am trying with<br/> graves now and forward attention enabled as well. Should I give up<br/> Tacotron altogether and try Taco2? I cannot figure out why it is<br/> overfitting, I thought it would surely do okay.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [March 20, 2020, 12:44pm]<br/> <br/> overfitting is not a huge deal with TTS. Just check the audio quality<br/> listening to the samples.<br/> <br/> When you use graves attention forward attention is out of service by<br/> default.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 20, 2020, 12:50pm]<br/> <br/> Ah okay, thanks, I had a suspicion. It's just that even after a day all<br/> I get in the testing samples is static noice and even teacher forcing in<br/> eval does not look great (plus the alignment score plateaus after 2k<br/> steps and does not improve at all).<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [March 23, 2020, 9:42am]<br/> <br/> if it is not the alignment, then there might be something else broken as<br/> well.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [March 24, 2020, 1:33am]<br/> <br/> I am<br/> really interested to see your results<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 24, 2020, 9:08am]<br/> <br/> Right, so,<br/> <br/> As I said, I ended up training (on the master branch) Tacotron on all<br/> the female speakers of Libri-TTS-360. At first I tried to use Forward<br/> Attention, but the only mechanism that worked was Graves. I tried<br/> bidirectional but sadly it didn't work (kept throwing memory errors). I<br/> also needed to disable `self.attention.init_win_idx()` on<br/> `layers/tacotron.py,` because otherwise it refused to synthesize. I am<br/> at approximately 103k steps now. What I have gathered:<br/> <br/> random intervals, the alignment just drops extremely low and then it<br/> goes back up, gradually.<br/> voices; however, punctuation like commas breaks it (it doesn't read<br/> further than the comma break). I am uploading synthesis with speaker<br/> 1 as the flag and speaker 14 as the flag.<br/> <br/> I don't have any test spectograms, because I just started retraining<br/> again. Again, my goal is more of producing a novel speaker using<br/> embeddings. I will let it train more and see how it goes, then use<br/> external embeddings I have extracted using the speaker encoder and feed<br/> these, instead of the speaker embedding layer. I would like to<br/> contribute the model I am training right now for the TTS project page on<br/> git, if it turns out to be good, along with the changes needed for<br/> loading your own embeddings.<br/> <br/> Any thoughts? I wonder if I can use this model to start a training<br/> session using forward attention, or batch normalization, after a certain<br/> number of steps. I also left r at 7, because gradual training is<br/> enabled, so I didn't think changing it would do anything.<br/> <br/> As always, thanks for all the work! Really happy if I can contribute in<br/> any way.<br/> <br/> 09.09.22.filename}[2880×1252 574<br/> <br/> *slash* <br/> <br/> 09.09.37.filename}[2880×1252 545<br/> <br/> *slash* <br/> <br/> 09.29.20.filename}[2300×880 516<br/> <br/> *slash* <br/> <br/> 09.29.27.filename}[2300×880 526<br/> <br/> *slash* <br/> [samples.zip](https://discourse.mozilla.org/uploads/short-url/4I9212401FZPG5SBdvaKf5b8CMx.zip)<br/> (104.2 KB)<br/> <br/> [](https://discourse.mozilla.org/t/failing-to-synthesize-when-using-gravesattention/59333)<br/> <br/> [Failing to synthesize when using GravesAttention<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 24, 2020, 9:58am]<br/> <br/> Update: Test figures have come in<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [March 24, 2020, 12:00pm]<br/> <br/> for a multi-speaker mode it takes longer for the mode to show a<br/> reasonable performance. I could only see it works good enough after 700K<br/> iterations. So maybe it is better to be patient. You might also train<br/> another model finetunning a pretrained LJSpeech model. That might make<br/> the problem easier.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 24, 2020, 12:03pm]<br/> <br/> I tried to do it on a pretrained LJSpeech model, but it said that 'as of<br/> now, you cannot introduce new speakers to an already trained model'.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [March 24, 2020, 12:03pm]<br/> <br/> I tried to do it on a pretrained LJSpeech model, but it said that 'as of<br/> now, you cannot introduce new speakers to an already trained model'.<br/> <br/> [**[next page<br/> →](https://discourse.mozilla.org/t/multispeaker-development-progress/55342?page=2)**<br/> <br/> [Archived Post]<br/>
